pkg/idutil/seq_generator.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package idutil
+
+import "sync"
+
+// SeqGenerator generates ID in sequential order.
+// It guarantees that generated ID is only used by the caller until it is recycled.
+// After using generated ID, caller should recycle the ID to SeqGenerator.
+type SeqGenerator struct {
+	mu   sync.Mutex
+	next int64              // next possible ID value
+	idm  map[int64]struct{} // ID map to track IDs in use
+}
+
+func NewSeqGenerator() *SeqGenerator {
+	return &SeqGenerator{
+		idm: make(map[int64]struct{}),
+	}
+}
+
+// Next generates a unique id.
+// The returned ID must be positive.
+func (g *SeqGenerator) Next() int64 {
+	for {
+		id := g.next
+		g.next++
+
+		// if id is still in use, skip it
+		if _, ok := g.idm[id]; ok {
+			continue
+		}
```

Comment:
`g.mu.Lock()`, `defer g.mu.Unlcok()` to be safe or is it only for sequential call?